gateway/listener.go

```diff
@@ -1,61 +1,47 @@
 package gateway
 
 import (
-	"errors"
 	"net"
 
 	"github.com/haveachin/infrared/connection"
 )
 
-var (
-	ErrCantStartOutListener = errors.New("failed to start outer listener")
-	ErrCantStartListener    = errors.New("failed to start listener")
-)
-
-type OuterListanerFactory func(addr string) OuterListener
-
-type OuterListener interface {
-	Start() error
-	Accept() (net.Conn, net.Addr)
-}
-
-func NewBasicOuterListener(addr string) OuterListener {
-	return &BasicOuterListener{addr: addr}
-}
 
-type BasicOuterListener struct {
-	net.Listener
-	addr string
-}
+type ErrorLogger func(err error)
+type ListanerFactory func(addr string) (net.Listener, error)
 
-func (l *BasicOuterListener) Start() error {
-	netL, err := net.Listen("tcp", l.addr)
-	if err != nil {
-		return ErrCantStartOutListener // TODO: look into ways to test this
+//The last argument is being used for an optional logger only the first logger will be used
+func NewBasicListener(listener net.Listener, ch connection.HandshakeChannel, errorLogger ...ErrorLogger) BasicListener {
+	logger := func(err error) {}
+	if len(errorLogger) != 0 {
+		logger = errorLogger[0]
+	}
```

Comment:
Thats almost correct, yes you could do something like this.
```
logger := func(err error) {}
for _, v := range errorLogger {
	logger = v
	break
}
```
If you dont use the first line, if the slice is nil, it will throw an panic bc logger is also nil when you try to run the method. and without the break the intended behavior will change from the first to the last logger. 
 
But why should you if this also works?